cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -163,27 +162,41 @@ public void createTethering(FullHttpRequest request, HttpResponder responder, @P
       throw e;
     }
 
-    // We don't need to keep track of the client metadata on the server side.
-    PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaceAllocations(), Collections.emptyMap(),
-                                                 tetherRequest.getDescription());
-    // We don't store the peer endpoint on the server side because the connection is initiated by the client.
-    PeerInfo peerInfo = new PeerInfo(peer, null, TetheringStatus.PENDING, peerMetadata,
-                                     tetherRequest.getRequestTime());
     try {
-      store.addPeer(peerInfo);
-    } catch (PeerAlreadyExistsException pae) {
-      // Peer is already configured, treat this as a no-op.
-      responder.sendStatus(HttpResponseStatus.OK);
-      return;
-    } catch (Exception e) {
+      String content = request.content().toString(StandardCharsets.UTF_8);
+      TetheringConnectionRequest tetherRequest = GSON.fromJson(content, TetheringConnectionRequest.class);
+
+      // We don't need to keep track of the client metadata on the server side.
+      PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaceAllocations(), Collections.emptyMap(),
+                                                   tetherRequest.getDescription());
+      try {
+        PeerInfo existingPeerInfo = store.getPeer(peer);
+        if (existingPeerInfo.getTetheringStatus() == TetheringStatus.REJECTED) {
+          // If peer exists in rejected state, send 403 to the client
+          responder.sendString(HttpResponseStatus.FORBIDDEN, TETHERING_REJECTED_MESSAGE);
+          return;
+        }
+        // Update the peer info. Keep the existing tethering status.
+        PeerInfo updatedPeerInfo = new PeerInfo(peer, null, existingPeerInfo.getTetheringStatus(),
+                                                peerMetadata, tetherRequest.getRequestTime(),
+                                                existingPeerInfo.getLastConnectionTime());
+        store.writePeer(updatedPeerInfo);
+        responder.sendStatus(HttpResponseStatus.OK);
+      } catch (PeerNotFoundException e) {
```

Comment:
this type of control logic should not happen at the handler layer. Everything needs to be handled within a single transaction, otherwise there are race conditions. More specifically, the handler should just be calling writePeer() and translating any exceptions into the right http response.

Reply:
moved logic into writePeerdf